server/bin/pbench-restore-and-unpack-tarballs.py

```diff
@@ -0,0 +1,243 @@
+            if c_entry.name.startswith(".") and c_entry.is_dir(follow_symlinks=False):
+                continue
+            if not c_entry.is_dir(follow_symlinks=False):
+                continue
+            # We have a controller directory.
+            with os.scandir(c_entry.path) as controller_scan:
+                for entry in controller_scan:
+                    if entry.name.startswith(".") and entry.is_dir(
+                        follow_symlinks=False
+                    ):
+                        continue
+                    if entry.is_dir(follow_symlinks=False):
+                        continue
+                    match = tb_pat.fullmatch(entry.name)
+                    if not match:
+                        continue
+                    # Turn the pattern components of the match into a datetime
+                    # object.
+                    tb_dt = datetime(
+                        int(match.group(1)),
+                        int(match.group(2)),
+                        int(match.group(3)),
+                        int(match.group(4)),
+                        int(match.group(5)),
+                        int(match.group(6)),
+                    )
+                    tb = Path(entry.path[:-4])
+                    if not tb.exists():
+                        raise Exception(
+                            f"Tar ball .md5, '{entry.path}', does not have"
+                            f" an associated tar ball, '{entry.path[:-4]}'"
+                        )
+                    yield tb_dt, c_entry.name, tb
+
+
+def main(options):
+    if not options.cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set"
+            " _PBENCH_SERVER_CONFIG env variable",
+            file=sys.stderr,
+        )
+        return 1
+
+    try:
+        config = pbench.PbenchConfig(options.cfg_name)
+    except pbench.BadConfig as e:
+        print(f"{_NAME_}: {e}", file=sys.stderr)
+        return 2
+
+    archive = config.ARCHIVE
+    archive_p = Path(archive).resolve(strict=True)
+
+    if not archive_p.is_dir():
+        print(
+            f"The configured ARCHIVE directory, {archive}, is not a valid" " directory",
+            file=sys.stderr,
+        )
+        return 4
+
+    backup = config.conf.get("pbench-server", "pbench-backup-dir")
+    backup_p = Path(backup).resolve(strict=True)
+
+    if not backup_p.is_dir():
+        print(
+            "The configured pbench-backup-dir directory, {backup}, is not a"
+            " valid directory",
+            file=sys.stderr,
+        )
+        return 6
+
+    start = pbench._time()
+
+    if options.dry_run:
+        print(f"Restoring tar balls from backup volume, {backup} (started at: {start})")
+
+    gen = gen_list(backup_p)
+    tbs_sorted = sorted(
+        [
+            (tb_dt, controller_name, tar_ball_name)
+            for tb_dt, controller_name, tar_ball_name in gen
+        ],
+        reverse=True,
+    )
```

Comment:
This code is quite specific to the `b0.69` branch.  Porting this forward to the `main` branch would only be done after we have a way to backup the ownership information of the tar ball.

As such, I'd like not to pursue too many more changes given the ability to verify this via tests.

Reply:
Since `gen` already yields tuples, there is no need to destructure its return only to re-compose it.  Something like this would do exactly what you code is doing now, in a much more readable way:
```
    tbs_sorted = sorted(list(gen), reverse=True)
```